### ISSUE TITLE
Can I train a custom dataset using YOLOv8-P2 on the command line?

### ISSUE THREAD

**Issue by Soichi9** [ultralytics/ultralytics#4084]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am working on small object detection. I saw the previous method in Python code (#981), but how do I do it on the command line? Should I change the contents of YOLOV8.yaml to YOLOv8-p2.yaml? 

The following is the code that you need to enter on the command line to run the training.The yolov8l.pt file is currently downloaded in the home directory.

`yolo detect train model=yolov8l.pt data=dataset.yaml device=1 epochs=400 batch=8 imgsz=640 degrees=360 fliplr=0.0`

**Comment by glenn-jocher** [comment]
@Soichi9 yes, you can train a custom dataset using YOLOv8-P2 on the command line. The command line arguments you've provided are almost correct, with one minor change: Instead of `model=yolov8l.pt`, you should specify the YAML configuration file for YOLOv8-P2, which might look something like `model=yolov8-p2.yaml`. YAML files are the correct way to specify the model structure for YOLOv8.

The rest of the command is generally correct, but remember that these options (`epochs`, `batch`, `imgsz`, etc.) should match the conditions of your custom dataset for optimal results. For example, `imgsz=640` specifies the training resolution, and your dataset images should be of comparable resolution. 

In case you are dealing with tiny object detection, you might also consider playing around with the `stride` parameter in the YAML configuration file, and adjust the `anchors` to the size of the object in your dataset.

Also, just a reminder that the `data=dataset.yaml` should point to your custom dataset configuration file. Make sure that the dataset paths and class labels in this YAML file correctly match with your actual dataset.

Finally, ensure that your YOLOv8-P2 model is properly set up and the needed weight files are available in your environment before starting the training process.

If you have further questions or encounter any issues in your training, feel free to ask. We’re here to help.

**Comment by Soichi9** [comment]
Thank you. The code worked successfully. Just to clarify, is this approach about training from scratch rather than fine-tuning a pre-trained model?

**Comment by herrig** [comment]
@glenn-jocher I am not sure if my training is correct, I have this code to train using yolov8-p2 configuration.

from ultralytics import YOLO

model = YOLO('yolov8x.pt')

model.train(data="/blue/nsboyd/re.herrigfurlane/Teste/Hole_v8/yolov8-p2.yaml", 
            workers=16,
            batch=64, 
            epochs=1200, 
            imgsz=640, 
            lr0=0.010, 
            lrf=0.005,
            seed=1234,
            momentum=0.900,)

#######################################

I configured my .yaml in this way, but I am not sure if the model is being trained using the correct structure.

train: /blue/nsboyd/re.herrigfurlane/Teste/row_final
val: /blue/nsboyd/re.herrigfurlane/Teste/row_final
test: /blue/nsboyd/re.herrigfurlane/Teste/row_final

nc: 3
names: ['Broadleaf', 'Grass', 'Nutsedge']

scales: # model compound scaling constants, i.e. 'model=yolov8n.yaml' will call yolov8.yaml with scale 'n'
  # [depth, width, max_channels]
  n: [0.33, 0.25, 1024]
  s: [0.33, 0.50, 1024]
  m: [0.67, 0.75, 768]
  l: [1.00, 1.00, 512]
  x: [1.00, 1.25, 512]

# YOLOv8.0 backbone
backbone:
  # [from, repeats, module, args]
  - [-1, 1, Conv, [64, 3, 2]]  # 0-P1/2
  - [-1, 1, Conv, [128, 3, 2]]  # 1-P2/4
  - [-1, 3, C2f, [128, True]]
  - [-1, 1, Conv, [256, 3, 2]]  # 3-P3/8
  - [-1, 6, C2f, [256, True]]
  - [-1, 1, Conv, [512, 3, 2]]  # 5-P4/16
  - [-1, 6, C2f, [512, True]]
  - [-1, 1, Conv, [1024, 3, 2]]  # 7-P5/32
  - [-1, 3, C2f, [1024, True]]
  - [-1, 1, SPPF, [1024, 5]]  # 9

# YOLOv8.0-p2 head
head:
  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 6], 1, Concat, [1]]  # cat backbone P4
  - [-1, 3, C2f, [512]]  # 12

  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 4], 1, Concat, [1]]  # cat backbone P3
  - [-1, 3, C2f, [256]]  # 15 (P3/8-small)

  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 2], 1, Concat, [1]]  # cat backbone P2
  - [-1, 3, C2f, [128]]  # 18 (P2/4-xsmall)

  - [-1, 1, Conv, [128, 3, 2]]
  - [[-1, 15], 1, Concat, [1]]  # cat head P3
  - [-1, 3, C2f, [256]]  # 21 (P3/8-small)

  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 12], 1, Concat, [1]]  # cat head P4
  - [-1, 3, C2f, [512]]  # 24 (P4/16-medium)

  - [-1, 1, Conv, [512, 3, 2]]
  - [[-1, 9], 1, Concat, [1]]  # cat head P5
  - [-1, 3, C2f, [1024]]  # 27 (P5/32-large)

  - [[18, 21, 24, 27], 1, Detect, [nc]]  # Detect(P2, P3, P4, P5)

**Comment by glenn-jocher** [comment]
@herrig yes, from your posted code and configurations, it looks like you are doing it correctly. You've specified your dataset paths for training, validation, and testing under 'train', 'val', and 'test' respectively in your YAML file, defined the number of classes 'nc' and their names right under with 'names', and provided the model architecture's details with 'backbone' and 'head' sections. Your YOLOv8-P2 structure seems to be correctly defined.

In terms of the python code shared, you've initialized the YOLO model with a YOLOv8x.pt as a pretrained architecture, and you're calling the `train()` function on this model to start the training process. The arguments you're passing to the train function, such as the path to your custom dataset's YAML file, the batch size, epochs, initial learning rate, etc., seem correct as well.

Once the code is executed, the model should start training with your specified configurations. You can monitor the training process via the logs or output, depending on your Python environment. If there is a problem during training, those logs will provide more visibility into the underlying issue.

Without running the code it's still hard to say with 100% certainty, but from an inspection of your posted Python and YAML content, it does appear to be set up correctly for training a YOLO model. Please monitor the output and logs during training for any potential issues that might arise.

**Comment by glenn-jocher** [comment]
@PTA00 yes, your approach is correct. By using 'yolov8n-p2.yaml' and 'yolov8n.pt', you are specifying the model configuration and pre-trained weights correctly, which should avoid any warnings related to model specification. Your training command also looks good, with the correct parameters for `data`, `epochs`, `batch`, and `imgsz`. Happy training! 🚀

**Comment by binhere** [comment]
Hello @glenn-jocher and @PTA00 ,  I am also working with `yolov8-p2.yaml`. But I do not know what the difference is between

`model = YOLO('yolov8n-p2.yaml').load('yolov8n.pt')` (1)

and

`model = YOLO('yolov8n-p2.yaml')` (2)

Question 1:
Does approach (1) transfer the `yolov8n.pt`'s weights to `yolov8n-p2.yaml` ? In other words, `yolov8n-p2.yaml` uses **a portion of the pre-trained weights of `yolov8n.pt`** on the COCO dataset?

Question 2:
How small of an object can the YOLOv8n-p2 model handle well?

**Comment by glenn-jocher** [comment]
@binhere (1) builds the P2 architecture and transfers any matching layers from `yolov8n.pt` (backbone/neck map; new P2-specific layers initialize randomly), while (2) starts from random initialization—this is the same “build from YAML and transfer weights” workflow shown in the training docs; see the build‑from‑YAML transfer example under Model Training and the train settings reference for tuning `imgsz` and related args. For tiny objects, the P2 (stride‑4) head boosts recall, but as a rule of thumb you’ll want targets ≥8–16 pixels on the short side at your chosen `imgsz`; if they’re smaller, increase `imgsz` or tile and compare `val` metrics across sizes. If you’re starting fresh, consider YOLO11n for better accuracy/speed at similar sizes. [Model Training (build from YAML and transfer weights)](https://docs.ultralytics.com/modes/train/), [Train settings (`imgsz`, etc.)](https://docs.ultralytics.com/usage/cfg/), [YOLO11 models overview](https://docs.ultralytics.com/models/yolo11/).
